core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/NearbyCompositeRuinMoveSelector.java

```diff
@@ -1,16 +1,18 @@
 package org.optaplanner.core.impl.heuristic.selector.move.generic;
 
 import java.util.Collection;
+import java.util.Collections;
 import java.util.Iterator;
 import java.util.List;
 
 import org.optaplanner.core.impl.domain.variable.descriptor.GenuineVariableDescriptor;
 import org.optaplanner.core.impl.domain.variable.descriptor.ShadowVariableDescriptor;
 import org.optaplanner.core.impl.heuristic.move.Move;
 import org.optaplanner.core.impl.heuristic.selector.entity.EntitySelector;
+import org.optaplanner.core.impl.heuristic.selector.entity.decorator.FilteringEntitySelector;
 
 public class NearbyCompositeRuinMoveSelector<Solution_> extends SimpleCompositeRuinMoveSelector<Solution_> {
-    private final EntitySelector<Solution_> rightEntitySelector;
+    private EntitySelector<Solution_> rightEntitySelector;
```

Comment:
I'm confused - why do you need to overwrite the selector?
If you only need the filtering selector locally inside `iterator`, create a local variable there.
If you need the filtering selector globally, why not store it in this field?

Reply:
The filtering selector seems to be a kind of wrapper over the basic selector. I cannot create it locally without having the base selector, that was set within the constructor.

Speaking of the second option: it's too early to do it in the constructor. The actual creation of the ruin move selector by the factory is before any solving is started. It is too early to figure out which entities are initialized and which not. We can do this only after we start iterating over the move selector (e.g. after some local search phase, before ruin). This is why it is done in the iterator()

Reply:
edit: I was wrong. For some reason I thought the filtering is done in FilteringEntitySelector constructor, which is not true. Applied